### PR TITLE
Add CollisionResult::nearest_points

### PR DESCRIPTION
**Pull request by jmirabel**
This supersedes #198 .

Work left to be done:
- [ ] agree on the API
- [x] handle shape-shape collision
- [ ] add some unit test
- [x] revise the distance lower bound (does not take the security margin).

Class ShapeCollisionTraversal should be removed.

As a bonus, this also solves #199.

### CODE REVIEW

**Comment by jmirabel** [comment]
@nim65s does it make sense to turn off travis (except for osx) ?

**Comment by nim65s** [comment]
I prefer not to have opinions about travis. If someone wants to use and maintain it, that's perfectly fine for me, but on my side I have enough work and results with one CI stack :)

**Comment by jmirabel** [comment]
After some effort, I ended up with the following.
- for collision queries, all the returned distances take the security margin into account,
- the collision result stores valid nearest_points whenever the distance lower bound is inferior to `break_distance`. When the distance lower bound is higher, their validity depends upon the pair of object type (maybe we should add a boolean that says whether they are valid).
- the distance lower bound can be negative, in which case it is the opposite of the penetration depth (the security margin is taken into account).

This leads to one strange thing. With the current changes, when the distance lower bound is inferior to the break distance, we don't have `||AB|| = distance_lower_bound` but ||AB|| = distance_lower_bound + security_margin` (A and B being the nearest point). i.e. the nearest point are not on the shapes inflated by `security_margin/2`

Some improvements, kept for later:
- For BVH, the minimal distance between two bounding volumes could be decrease as the algo finds smaller distance lower bound. This is especially useful when the `break_distance` is set to a high value.
- Rename `break_distance` to something easier to understand from the user perspective.
- Clean API of CollisionTraversalNode:
  - remove ShapeCollisionTraversalNode,
  - merge the two CollisionTraversalNode::BVDisjoint prototype into one. The squared distance lower bound attribute is not useless.
- The above API should be enforced by some unit-tests.

**Comment by jmirabel** [comment]
This PR is ready for reviews.

File: src/collision.cpp [hunk 1]
```diff
@@ -61,6 +61,7 @@ void CollisionResult::swapObjects()
   {
     std::swap(it->o1, it->o2);
     std::swap(it->b1, it->b2);
+    it->normal *= -1;
```
Comment (jcarpent) [inline src/collision.cpp:64]: This is a critical bug, isn't ?

**Review by jcarpent — APPROVED** [review]

File: src/collision.cpp (hunk 1 above)
Comment (jmirabel) [reply src/collision.cpp:64]: Yes. I am surprised it hasn't popup earlier.

File: include/hpp/fcl/internal/traversal_node_bvh_shape.h [hunk 2]
```diff
@@ -337,27 +346,32 @@ class ShapeMeshCollisionTraversalNode : public ShapeBVHCollisionTraversalNode<S,
...
-    }
-    assert (!this->result->isCollision () || sqrDistLowerBound > 0);
+    } else if (distToCollision < 0) {
+      sqrDistLowerBound = 0;
+      if(this->request.num_max_contacts > this->result->numContacts())
+      {
+        this->result->addContact(Contact(this->model1, this->model2,
+              Contact::NONE, primitive_id,
+              .5 * (c1+c2), (c2-c1).normalized (),
+              -distance));
```
Comment (jmirabel) [inline include/hpp/fcl/internal/traversal_node_bvh_shape.h:366]: TODO: sometimes the contact distance takes the security margin into account. Sometimes it does not. It should be homogenized.

File: src/collision_func_matrix.cpp [hunk 3]
```diff
@@ -82,34 +82,20 @@ std::size_t ShapeShapeCollide(const CollisionGeometry* o1, const Transform3f& tf
...
+  const Vec3f& p1 = distanceResult.nearest_points [0];
+  const Vec3f& p2 = distanceResult.nearest_points [1];
+  FCL_REAL distToCollision = distance - request.security_margin;
 
-      Contact contact (o1, o2, distanceResult.b1, distanceResult.b2,
-          (p1+p2)/2,
-          distanceResult.normal,
-          -distance+request.security_margin);
+  internal::updateDistanceLowerBoundFromLeaf (request, result, distToCollision, p1, p2);
+  if (distToCollision <= 0 && result.numContacts () < request.num_max_contacts) {
```
Comment (jcarpent) [inline src/collision_func_matrix.cpp:90]: Currently, this results is not taking into account the break_distance information.
Compared to other collision checks.
```suggestion
  if (distToCollision <= request.break_distance && result.numContacts () < request.num_max_contacts) {
```

**Review by jcarpent — CHANGES_REQUESTED** [review]
Maybe, this should fix another bug

File: src/collision_func_matrix.cpp (hunk 3 above)
Comment (jmirabel) [reply src/collision_func_matrix.cpp:90]: No, there is a contact only if distToCollision <= 0. At the moment, `break_distance` is not useful for shape-shape pairs.

File: include/hpp/fcl/internal/traversal_node_bvh_shape.h (hunk 2 above)
Comment (jcarpent) [reply include/hpp/fcl/internal/traversal_node_bvh_shape.h:366]: Why to not apply the change here?

Comment (jmirabel) [reply include/hpp/fcl/internal/traversal_node_bvh_shape.h:366]: I plan to do it. I wrote it so as not to forget.

**Comment by jcarpent** [comment]
@jmirabel Could you consider adjusting this PR for merging it?

**Comment by jmirabel** [comment]
Not sure what you are asking exactly. I can definitely resolve the git conflicts quickly. Adding unit test, finalizing the code and maybe update the API is another story. I know I won't have time before beginning of March. Then I am not sure yet.

**Comment by jmirabel** [comment]
I assigned myself so that I see this in my list of assigned PRs.

File: src/collision_node.cpp [hunk 4]
```diff
@@ -60,7 +60,13 @@ void collide(CollisionTraversalNodeBase* node,
       collisionRecurse(node, 0, 0, front_list, sqrDistLowerBound);
     else
       collisionNonRecurse(node, front_list, sqrDistLowerBound);
-    result.updateDistanceLowerBound (sqrt (sqrDistLowerBound));
```
Comment (jcarpent) [inline src/collision_node.cpp:63]: @jmirabel Why do you remove that here?

**Comment by jcarpent** [comment]
Supersedes by #303.

**Pull request closed without merging**